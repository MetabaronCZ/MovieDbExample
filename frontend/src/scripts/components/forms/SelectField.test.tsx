import { act } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';
import {
  fireEvent,
  getAllByRole,
  getByRole,
  getByText,
  queryByRole,
} from '@testing-library/dom';

import { SelectField } from 'components/forms/SelectField';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

describe('components/forms/SelectField', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <SelectField
        id="ID"
        label="LABEL"
        value={1}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        required
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const label = getByText(container, 'LABEL');
    expect(label).toBeInTheDocument();
    expect(label.tagName.toLowerCase()).toEqual('label');
    expect(label).toHaveAttribute('for', 'ID');

    const required = label.querySelector('span');
    expect(required).toBeInTheDocument();

    const select = getByRole(container, 'button');
    expect(select).toHaveAttribute('id', 'ID');
    expect(select).toHaveTextContent('Item 1');
  });

  it('should render multi-select variant', async () => {
    const { container } = await waitForComponent(
      <SelectField
        id="ID"
        label="LABEL"
        value={[]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        multi
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const label = getByText(container, 'LABEL');
    expect(label).toBeInTheDocument();

    const select = getByRole(container, 'button');
    expect(select).toHaveAttribute('id', 'ID');
    expect(select).toHaveTextContent('- vyberte -');
  });

  it('should render error', async () => {
    const { container } = await waitForComponent(
      <SelectField
        label="LABEL"
        error="SOME-ERROR"
        value="VALUE"
        options={[]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const error = getByText(container, 'SOME-ERROR');
    expect(error).toBeInTheDocument();
  });

  it('should render info text', async () => {
    const { container } = await waitForComponent(
      <SelectField
        label="LABEL"
        info="SOME-INFO"
        value="VALUE"
        options={[]}
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const info = getByText(container, 'SOME-INFO');
    expect(info).toBeInTheDocument();
  });

  it('should be able to change value', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectField
        label="LABEL"
        value={1}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={onSelect}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const handle = getByRole(container, 'button');
    expect(handle).toBeInTheDocument();

    let list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();

    // open Select by handle click
    act(() => {
      fireEvent.click(handle);
    });

    list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const items = getAllByRole(list, 'button');
    expect(items.length).toEqual(2);

    // click Item 2
    act(() => {
      fireEvent.click(items[1]);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith(2); // Item 2 value

    // Select is closed after selection
    list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();
  });

  it('should not open when disabled', async () => {
    const { container } = await waitForComponent(
      <SelectField
        label="LABEL"
        value={1}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        disabled
        onSelect={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    const handle = getByRole(container, 'button');
    expect(handle).toBeInTheDocument();

    let list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();

    // try open Select by handle click
    act(() => {
      fireEvent.click(handle);
    });

    list = queryByRole(container, 'list');
    expect(list).not.toBeInTheDocument();
  });
});
