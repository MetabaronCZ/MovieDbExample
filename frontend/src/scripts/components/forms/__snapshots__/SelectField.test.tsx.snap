// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/forms/SelectField should render error 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: row;
  gap: 8px;
  align-items: center;
  justify-content: start;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  color: #f00;
}

<div>
  <fieldset
    class="c0"
    data-invalid="true"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="select-0"
      >
        LABEL
      </label>
    </div>
    <p
      class="c3"
    >
      SOME-ERROR
    </p>
  </fieldset>
</div>
`;

exports[`components/forms/SelectField should render info text 1`] = `
.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: row;
  gap: 8px;
  align-items: center;
  justify-content: start;
}

.c3 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  color: #98c3d9;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="select-1"
      >
        LABEL
      </label>
    </div>
    <p
      class="c3"
    >
      SOME-INFO
    </p>
  </fieldset>
</div>
`;

exports[`components/forms/SelectField should render multi-select variant 1`] = `
.c4 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c4:disabled {
  cursor: not-allowed;
}

.c6 {
  flex: 1;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  gap: 12px;
  line-height: 1;
  font-weight: 700;
  height: 40px;
  padding: 0 16px;
  padding-left: 12px;
  padding-right: 12px;
  border: 2px solid #002d56;
  background: #fff;
  width: 100%;
  height: 40px;
  gap: 8px;
  padding-left: 12px;
  padding-right: 12px;
}

.c5:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c5:focus {
  outline: 1px dashed #002d56;
}

.c5:disabled {
  color: #98c3d9;
  border-color: #98c3d9;
  background-color: #fff;
}

.c7 {
  font-size: 14px;
  font-weight: 400;
  text-align: left;
}

.c3 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: row;
  gap: 8px;
  align-items: center;
  justify-content: start;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="ID"
      >
        LABEL
      </label>
      <div
        class="c3"
      >
        <button
          class="c4 c5"
          id="ID"
          title="Otevřít"
          type="button"
        >
          <span
            class="c6 c7"
          >
            - vyberte -
          </span>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-down fa-lg "
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`components/forms/SelectField should render properly 1`] = `
.c5 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c5:disabled {
  cursor: not-allowed;
}

.c7 {
  flex: 1;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.c6 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  gap: 12px;
  line-height: 1;
  font-weight: 700;
  height: 40px;
  padding: 0 16px;
  padding-left: 12px;
  padding-right: 12px;
  border: 2px solid #002d56;
  background: #fff;
  width: 100%;
  height: 40px;
  gap: 8px;
  padding-left: 12px;
  padding-right: 12px;
}

.c6:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c6:focus {
  outline: 1px dashed #002d56;
}

.c6:disabled {
  color: #98c3d9;
  border-color: #98c3d9;
  background-color: #fff;
}

.c8 {
  font-size: 14px;
  font-weight: 400;
  text-align: left;
}

.c4 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c2 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: flex;
  flex-direction: row;
  gap: 4px;
  white-space: nowrap;
}

.c3 {
  color: #f00;
}

.c3::after {
  content: '*';
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 4px;
  border: none;
}

.c1 {
  position: relative;
  display: flex;
  flex-direction: row;
  gap: 8px;
  align-items: center;
  justify-content: start;
}

<div>
  <fieldset
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2 "
        for="ID"
      >
        <span
          class="c3"
        />
        LABEL
      </label>
      <div
        class="c4"
      >
        <button
          class="c5 c6"
          id="ID"
          title="Otevřít"
          type="button"
        >
          <span
            class="c7 c8"
          >
            Item 1
          </span>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angle-down fa-lg "
            data-icon="angle-down"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </fieldset>
</div>
`;
